corehq/apps/hqmedia/tasks.py

```diff
@@ -149,81 +137,126 @@ def build_application_zip(include_multimedia_files, include_index_files, app,
+            if len(media_suites) != 1:
+                message = _('Could not identify media_suite.xml in CCZ')
+                errors.append(message)
+            else:
+                with z.open(media_suites[0]) as media_suite:
+                    from corehq.apps.app_manager.xform import parse_xml
+                    parsed = parse_xml(media_suite.read())
+                    resources = {node.text for node in
+                                 parsed.findall("media/resource/location[@authority='local']")}
+                    names = z.namelist()
+                    missing = [r for r in resources if re.sub(r'^\.\/', '', r) not in names]
+                    errors += [_('Media file missing from CCZ: {}').format(r) for r in missing]
+
+
+def _build_ccz_files(build, build_profile_id, include_multimedia_files, include_index_files,
+                     download_id, compress_zip, filename, download_targeted_version):
+    from corehq.apps.hqmedia.views import iter_app_files
+    files, errors, file_count = iter_app_files(
+        build, include_multimedia_files, include_index_files, build_profile_id,
+        download_targeted_version=download_targeted_version,
+    )
+
+    if toggles.CAUTIOUS_MULTIMEDIA.enabled(build.domain):
+        manifest = json.dumps({
+            'include_multimedia_files': include_multimedia_files,
+            'include_index_files': include_index_files,
+            'download_id': download_id,
+            'build_profile_id': build_profile_id,
+            'compress_zip': compress_zip,
+            'filename': filename,
+            'download_targeted_version': download_targeted_version,
+            'app': build.to_json(),
+        }, indent=4)
+        files = itertools.chain(files, [('manifest.json', manifest)])
+    return files, errors, file_count
+
+
+def _zip_ccs_files(fpath, files, current_progress, file_progress, file_count, compression, track_progress):
+    with open(fpath, 'wb') as tmp:
+        with zipfile.ZipFile(tmp, "w") as z:
+            for path, data in files:
+                # don't compress multimedia files
+                extension = os.path.splitext(path)[1]
+                file_compression = zipfile.ZIP_STORED if extension in MULTIMEDIA_EXTENSIONS else compression
+                z.writestr(path, data, file_compression)
+                current_progress += file_progress / file_count
+                if track_progress:
+                    DownloadBase.set_progress(build_application_zip, current_progress, 100)
+
+
+def create_ccz_files(build, build_profile_id, include_multimedia_files=True, include_index_files=True,
+                     download_id=None, compress_zip=False, filename="commcare.zip",
+                     download_targeted_version=False, track_progress=False, expose_link=False):
+    """
+    :param track_progress: track progress of the files when being run asynchronously by celery
```

Comment:
I am scared about making changes to this code. I just kept it in the same order and specifics as it was earlier.

Reply:
splitting it up into multiple functions changes the equation a bit though - previously, only code inside `build_application_zip` referenced `build_application_zip`.  Now `create_ccz_files` and `zip_ccs_files` both reference that function, meaning that they're coupled to it. 

`build_application_zip` is being passed to `set_progress` as the way to communicate to celery what task is being processed.  For instance, how does `create_ccz_files` know that it's being called from `build_application_zip`?

Reply:
@esoergel 
i am not fully sure how the tracking of the task would change with these changes. I just tested it once and saw the progress bar updated but i can't say it worked exactly the way its suppose to
Are you saying the tracking would not work correctly now?

Reply:
Agree with @esoergel that instead of passing a boolean you should just pass the task function itself in which case you just change:
```diff
- if track_progress:
-     DownloadBase.set_progress(build_application_zip, current_progress, 100)
DownloadBase.set_progress(celery_task, current_progress, 100)
```

You don't even have to check if it's `None` since that's handled in `set_progress`